src/core/composer/qgsaddremoveitemcommand.cpp

```diff
@@ -80,5 +80,5 @@ void QgsAddRemoveItemCommand::switchState()
     emit itemAdded( mItem );
     mState = Added;
   }
-  QgsProject::instance()->setDirty( true );
+  mComposition->project()->setDirty( true );
```

Comment:
Yep... There's nothing in this API forcing the project use, so happy to say project is mandatory for now and address this when required!